kv/kvserver: TestReplicaRangefeedErrors failed

kv/kvserver.TestReplicaRangefeedErrors [failed](https://tanzanite.cluster.engflow.com/invocations/default/c687761b-7cee-47eb-a5cf-bcb6a4153377?testReportRun=7&testReportShard=28&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [a23be6bce928c3e08074a15815b3c67a657bb40e](https://github.com/cockroachdb/cockroach/commits/a23be6bce928c3e08074a15815b3c67a657bb40e):


```
=== RUN   TestReplicaRangefeedErrors
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReplicaRangefeedErrors2633960710
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_rangefeed_test.go:1278: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicaRangefeedErrors2633960710
--- FAIL: TestReplicaRangefeedErrors (45.95s)
=== RUN   TestReplicaRangefeedErrors/REASON_REPLICA_REMOVED
    replica_rangefeed_test.go:870: incorrect events on stream, found [checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<> >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733395375773891722 > >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733395375774325982 > > ], want [checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<> >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733395375773891722 > > ]
    --- FAIL: TestReplicaRangefeedErrors/REASON_REPLICA_REMOVED (6.22s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=28</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaRangefeedErrors.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45259

kv/kvserver.TestReplicaRangefeedErrors [failed](https://tanzanite.cluster.engflow.com/invocations/default/a3b78e3b-0ad5-495b-ad9a-a9e70ce6adfa?testReportRun=3&testReportShard=28&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [fa9acaa8a8b8b58b1448c61ca5a8f6e6a992d64c](https://github.com/cockroachdb/cockroach/commits/fa9acaa8a8b8b58b1448c61ca5a8f6e6a992d64c):


```
=== RUN   TestReplicaRangefeedErrors
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReplicaRangefeedErrors45091288
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_rangefeed_test.go:1278: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicaRangefeedErrors45091288
--- FAIL: TestReplicaRangefeedErrors (46.95s)
=== RUN   TestReplicaRangefeedErrors/REASON_RAFT_SNAPSHOT
    replica_rangefeed_test.go:1024: incorrect events on stream, found [checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<> >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733473152983858546 > >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733473152984320666 > > ], want [checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<> >  checkpoint:<span:<key:"\372a" end_key:"\372z" > resolved_ts:<wall_time:1733473152983858546 > > ]
    --- FAIL: TestReplicaRangefeedErrors/REASON_RAFT_SNAPSHOT (6.20s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=28</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaRangefeedErrors.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/replica_rangefeed_test.go
+++ b/pkg/kv/kvserver/replica_rangefeed_test.go
@@ -7,7 +7,6 @@ package kvserver_test
 
 import (
 	"context"
-	"reflect"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -779,17 +778,6 @@ func TestReplicaRangefeedErrors(t *testing.T) {
 		t *testing.T, stream *testStream, streamErrC <-chan error, span roachpb.Span,
 	) {
 		t.Helper()
-		noResolveTimestampEvent := kvpb.RangeFeedEvent{
-			Checkpoint: &kvpb.RangeFeedCheckpoint{
-				Span:       span,
-				ResolvedTS: hlc.Timestamp{},
-			},
-		}
-		resolveTimestampEvent := kvpb.RangeFeedEvent{
-			Checkpoint: &kvpb.RangeFeedCheckpoint{
-				Span: span,
-			},
-		}
 		var events []*kvpb.RangeFeedEvent
 		testutils.SucceedsSoon(t, func() error {
 			if len(streamErrC) > 0 {
@@ -805,21 +793,14 @@ func TestReplicaRangefeedErrors(t *testing.T) {
 		if len(streamErrC) > 0 {
 			t.Fatalf("unexpected error from stream: %v", <-streamErrC)
 		}
-		expEvents := []*kvpb.RangeFeedEvent{&noResolveTimestampEvent}
-		if len(events) > 1 {
-			// Unfortunately there is a timing issue here and the range feed may
-			// publish two checkpoints, one with a resolvedTs and one without, so we
-			// check for either case.
-			resolveTimestampEvent.Checkpoint.ResolvedTS = events[1].Checkpoint.ResolvedTS
-			expEvents = []*kvpb.RangeFeedEvent{
-				&noResolveTimestampEvent,
-				&resolveTimestampEvent,
-			}
-		}
-		if !reflect.DeepEqual(events, expEvents) {
-			t.Fatalf("incorrect events on stream, found %v, want %v", events, expEvents)
-		}
 
+		var lastTS hlc.Timestamp
+		for _, evt := range events {
+			require.NotNil(t, evt.Checkpoint, "expected only checkpoint events")
+			require.Equal(t, span, evt.Checkpoint.Span)
+			require.True(t, evt.Checkpoint.ResolvedTS.After(lastTS) || evt.Checkpoint.ResolvedTS.Equal(lastTS), "unexpected resolved timestamp regression")
+			lastTS = evt.Checkpoint.ResolvedTS
+		}
 	}
 
 	assertRangefeedRetryErr := func(

